Add --no-data-gathering as a startup option

s9ferech opened the pull request:
### Pull Request Description
<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->
This solves issue #1303. The option will disable sending any data to Mixpanel.

### Important Notes
<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->
There will still be a request to Mixpanel during initialisation (src/js/lib/content/index.js, line 175). No user specific data is transmitted during this initialisation, but we could circumvent it as well, if it it a problem.

### Checklist
Please include the following checklist in your PR:

- [x] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [ ] ~The documentation has been updated if necessary.~
- [ ] ~All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.~
- [ ] ~All code has automatic tests where possible.~
- [ ] ~All code has been profiled where possible.~
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [ ] ~All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).~
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

s9ferech committed:
Add a short description to src/js/config.js

s9ferech committed:
Add --no-data-gathering as a startup option


This option will disable sending any data to Mixpanel.

s9ferech committed:
Restore `package.json` for log server

We need this to test and run the log server. It was removed as part of
PR #978, probably accident, since I can't find a reason for that change.

s9ferech committed:
Include --no-data-gathering in changelog

s9ferech committed:
Fix formatting in CHANGELOG.md

wdanilo commented on src/js/config.js:
```diff
@@ -1,2 +1,6 @@
+/**
+ * Configuration options that are only used by the JavaScript part of the project
```
Docs need to be sentences - missing dot.

wdanilo commented on src/js/lib/client/src/index.js:
```diff
@@ -172,7 +172,12 @@ optParser.options('crash-report-host', {
+})
+
+optParser.options('no-data-gathering', {
+    describe    : 'Disable the sharing of any usage data with the Enso developers',
```
I would name it "Disable anonymous data collection" - it's shorter, a more popular name for that, and also doesn't mention "Enso" (we have a rule that we do not use product name too often if that's possible in order to keep the messages shorter + code easier to refactor).

MichaelMauderer replied:
"Disable anonymous data collection" could be misleading, though. It could imply that non-anonymous data gathering is happening.

s9ferech replied:
I changed it to "Disable anonymous data collection". I assume it will be clear enough.

wdanilo replied:
@MichaelMauderer oh good point. @s9ferech please confirm it with @iamrecursion - she always knows how to write sentences that are not ambiguous (maybe this one is correct, just confirm it with Ara please)

iamrecursion replied:
No. That one is definitely ambiguous. I would go with the one that is here originally.

wdanilo commented on src/js/lib/content/src/index.js:
```diff
@@ -450,18 +429,28 @@ function ok(value) {
 
-    //initCrashHandling()
+    API.remoteLog = config.noDataGathering
+                    ? function () {}
```
create var `let remoteLogDisabled = () => {}` instead the line above. We are trying to minimize the number of multi-line expressions in the codebase

wdanilo commented on src/js/lib/content/src/index.js:
```diff
@@ -450,18 +429,28 @@ function ok(value) {
+    API.remoteLog = config.noDataGathering
+                    ? function () {}
+                    : remoteLog
+    window.setInterval(() =>{API.remoteLog("alive");}, 1000 * 60)
```
please expose the `60 seconds` as a constant variable on top of the file (in constants section)

wdanilo commented on src/js/lib/log-server/package.json:
```diff
@@ -0,0 +1,31 @@
+      "mocha": "^8.2.1",
+      "mock-fs": "^4.13.0"
+    }
+  }
```
no nl

wdanilo replied:
no nl

wdanilo commented on src/js/lib/log-server/package.json:
```diff
@@ -0,0 +1,31 @@
+{
```
I don't understand this change. Why this file is added now / why it was missing before? This PR doesn't seem to have any connection to this file. (This question is the only blocker from merging it to me)

s9ferech replied:
Indeed, the file is not directly related to the purpose of this PR. I added it, so I can test if the crash reporting still works.

The file existed before but was deleted by accident, as far as I can tell. Or it was deleted with a good reason that I currently don't understand. I am not very familiar with Lerna. Is it correct that this package should have its own `package.json`? Or should that configuration go somewhere else?

wdanilo replied:
Im surprised the file was not there tbh. Could you please confirm it works before merging this PR? I mean - could you confirm this server compiles and runs based on this file config?

s9ferech replied:
I was also surprised. Maybe it got lost because we have `package.json` in our `.gitignore`.

The server does run with that config.

wdanilo reviewed (changes requested):
I have one blocker question, beside that, it looks fine. I'd feel better if also @MichaelMauderer would review the logic, as he was implementing it, and maybe there are some things that I overlooked.

MichaelMauderer reviewed (commented):
> There will still be a request to Mixpanel during initialisation (src/js/lib/content/index.js, line 175). No user specific data is transmitted during this initialisation, but we could circumvent it as well, if it it a problem.

I think it would be great if this could also disable the initial connection to mixpanel. That would ensure that not outgoing connection is required when using this cmd option. Doing an initial connection can delay the IDE startup in no- or low-connectivity environments.

s9ferech committed:
Address comments on PR #1328

Including
- New doc for --no-data-gathering flag
- Only initialise Mixpanel if data will be gathered

s9ferech committed:
Update log server's README

s9ferech committed:
Make documentation less ambiguous

s9ferech committed:
Merge branch 'develop' into wip/fr/no-data-gathering
